tpmutil/encoding.go

```diff
@@ -120,81 +55,96 @@ func packWithHeader(ch commandHeader, cmd ...interface{}) ([]byte, error) {
 // prepended length, to match how the TPM encodes variable-length arrays. If
 // you wish to add a byte slice without length prefix, use RawBytes.
 func Pack(elts ...interface{}) ([]byte, error) {
-	if lengthPrefixSize == 0 {
-		return nil, errors.New("lengthPrefixSize must be initialized")
-	}
-
 	buf := new(bytes.Buffer)
 	if err := packType(buf, elts...); err != nil {
 		return nil, err
 	}
-
 	return buf.Bytes(), nil
 }
 
-// packType recursively packs types the same way that encoding/binary does
-// under binary.BigEndian, but with one difference: it packs a byte slice as a
-// lengthPrefixSize size followed by the bytes. The function unpackType
-// performs the inverse operation of unpacking slices stored in this manner and
-// using encoding/binary for everything else.
-func packType(buf io.Writer, elts ...interface{}) error {
-	for _, e := range elts {
-		marshaler, ok := e.(SelfMarshaler)
-		if ok {
-			if err := marshaler.TPMMarshal(buf); err != nil {
-				return err
-			}
-			continue
+// tryMarshal attempts to use a Marshal() method defined on the type
+// to pack v into buf. True is returned if the method exists and the
+// marshal was attempted.
+func tryMarshal(buf io.Writer, v reflect.Value) (bool, error) {
+	t := v.Type()
+	if t.Implements(selfMarshalerType) {
+		return true, v.Interface().(SelfMarshaler).TPMMarshal(buf)
+	}
+
+	// We might have a non-pointer struct field, but we dont have a
+	// pointer with which to implement the interface.
+	// If the pointer of the type implements the interface, we should be
+	// able to construct a value to call TPMUnmarshal() with.
+	if reflect.PtrTo(t).Implements(selfMarshalerType) {
```

Comment:
Use `v.CanAddr()` and `v.Addr()` to call the pointer method

Reply:
No, that wont work if v is not addressable. This is checking if a pointer to the type of `v` would implement the interface, so we can construct an equivalent addressable type and then overwrite the reflect.Value.

This is needed in particular for literals with non-pointer fields that implement the interface.

Reply:
I know, but if v is not addressable, can it be settable?
Here's where this fails:

```
type Foo struct {
  BeFancy bool
  Val string
}

func (f *Foo) TPMUnmarshal(in bytes.Buffer) error {
  // read size header
  // read bytes based on size header into val
  if f.BeFancy {
    val = fmt.Sprintf("fancy(%s)", val)
  }
  return nil
}
```

With your current implementation, `BeFancy` will always be false.
But what if caller set it to true, expecting special unmarshaling?

Reply:
Yes I think you can have settable but not addressable. I originally tried that and many different permutations of what you are describing - they did not work in all cases. I created the logic to generate a surrogate type and assign it as a last resort.

I think the example you gave should work fine because of `tmp.Elem().Set(v)`.

If you passed that by pointer, it would work without hitting the logic of generating a new type. If it was passed by value, it would not work regardless.